src/go/rdctl/pkg/snapshot/manager.go

```diff
@@ -0,0 +1,286 @@
+package snapshot
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"regexp"
+	"time"
+
+	"github.com/rancher-sandbox/rancher-desktop/src/go/rdctl/pkg/paths"
+)
+
+var nameRegexp = *regexp.MustCompile("^[0-9a-zA-Z_-]{0,100}$")
+var ErrNameExists = errors.New("name already exists")
+var ErrInvalidName = fmt.Errorf("name does not match regex %q", nameRegexp.String())
+
+// Handles all snapshot-related functionality.
+type Manager struct {
+	Paths paths.Paths
+}
+
+// Represents a file that is included in a snapshot.
+type snapshotFile struct {
+	// The path that Rancher Desktop uses.
+	WorkingPath string
+	// The path that the file is backed up to before attempting
+	// a Restore.
+	BackupPath string
+	// The path that the file is put at in a snapshot.
+	SnapshotPath string
+	// Whether clonefile (macOS) or ioctl_ficlone (Linux) should be used
+	// when copying the file around.
+	CopyOnWrite bool
+	// Whether it is ok for the file to not be present.
+	MissingOk bool
+	// The permissions the file should have.
+	FileMode os.FileMode
+}
+
+func NewManager(paths paths.Paths) Manager {
+	return Manager{
+		Paths: paths,
+	}
+}
+
+func (manager Manager) getSnapshotFiles(id string) []snapshotFile {
+	snapshotDir := filepath.Join(manager.Paths.Snapshots, id)
+	files := []snapshotFile{
+		{
+			WorkingPath:  filepath.Join(manager.Paths.Config, "settings.json"),
+			SnapshotPath: filepath.Join(snapshotDir, "settings.json"),
+			CopyOnWrite:  false,
+			MissingOk:    false,
+			FileMode:     0o644,
+		},
+		{
+			WorkingPath:  filepath.Join(manager.Paths.Lima, "_config", "override.yaml"),
+			SnapshotPath: filepath.Join(snapshotDir, "override.yaml"),
+			CopyOnWrite:  false,
+			MissingOk:    true,
+			FileMode:     0o644,
+		},
+		{
+			WorkingPath:  filepath.Join(manager.Paths.Lima, "0", "basedisk"),
+			SnapshotPath: filepath.Join(snapshotDir, "basedisk"),
+			CopyOnWrite:  true,
+			MissingOk:    false,
+			FileMode:     0o644,
+		},
+		{
+			WorkingPath:  filepath.Join(manager.Paths.Lima, "0", "diffdisk"),
+			SnapshotPath: filepath.Join(snapshotDir, "diffdisk"),
+			CopyOnWrite:  true,
+			MissingOk:    false,
+			FileMode:     0o644,
+		},
+		{
+			WorkingPath:  filepath.Join(manager.Paths.Lima, "_config", "user"),
+			SnapshotPath: filepath.Join(snapshotDir, "user"),
+			CopyOnWrite:  false,
+			MissingOk:    false,
+			FileMode:     0o600,
+		},
+		{
+			WorkingPath:  filepath.Join(manager.Paths.Lima, "_config", "user.pub"),
+			SnapshotPath: filepath.Join(snapshotDir, "user.pub"),
+			CopyOnWrite:  false,
+			MissingOk:    false,
+			FileMode:     0o644,
+		},
+	}
+	for i := range files {
+		files[i].BackupPath = fmt.Sprintf("%s.backup", files[i].WorkingPath)
+	}
+	return files
+}
+
+// Creates a new snapshot.
+func (manager Manager) Create(name string) (Snapshot, error) {
```

Comment:
I think this would be better returning `(*Snapshot, error)` and then you can return `nil, err` on error, and `&snapshot, nil` at line 131.

Because go does automatic dereferencing of pointers no other changes would be necessary.